tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -3402,15 +3402,40 @@ SpirvEmitter::processACSBufferAppendConsume(const CXXMemberCallExpr *expr) {
   if (isAppend) {
     // Write out the value
     auto *arg0 = doExpr(expr->getArg(0));
+    if (!arg0)
+      return nullptr;
+
     if (!arg0->isRValue()) {
       arg0 = spvBuilder.createLoad(bufferElemTy, arg0);
     }
     storeValue(bufferInfo, arg0, bufferElemTy);
     return 0;
   } else {
-    // Note that we are returning a pointer (lvalue) here inorder to further
-    // acess the fields in this element, e.g., buffer.Consume().a.b. So we
-    // cannot forcefully set all normal function calls as returning rvalue.
+    // Note that we are returning a rvalue here. If it is a pointer, when
+    // there is a type-casting, it causes an error e.g.,
+    //
+    // AppendStructuredBuffer<int> foo;
+    // ConsumeStructuredBuffer<bool> bar;
+    // foo.Append(bar.Consume());
+    //
+    // In addition, we type-cast the rvalue to bool type rvalue if it must
+    // be a bool type because its caller expects its return type is a bool
+    // type. For example, the above code has an error without type-casting
+    // because of an implicit type-cast from bool to int.
+    if (!bufferInfo->isRValue()) {
+      bufferInfo = spvBuilder.createLoad(bufferElemTy, bufferInfo);
+
+      if (bufferInfo->getLayoutRule() != SpirvLayoutRule::Void &&
+          isBoolOrVecMatOfBoolType(bufferElemTy)) {
```

Comment:
If I remember correctly, we don't support boolean matrix at this point. So you should use `isBoolOrVecOfBoolType` instead of `isBoolOrVecMatOfBoolType`.
Also the body of the if-condition is assuming it's a vector or a scalar bool (not matrix).